llms/mlx_lm/models/mamba.py

```diff
@@ -0,0 +1,259 @@
+        self.ssm_state_size = args.state_size
+        self.conv_kernel_size = args.conv_kernel
+        self.intermediate_size = args.intermediate_size
+        self.time_step_rank = int(args.time_step_rank)
+        self.use_conv_bias = args.use_conv_bias
+
+        self.in_proj = nn.Linear(self.hidden_size, self.intermediate_size * 2, bias=args.use_bias)
+
+        self.conv1d = DepthWiseConv1d(
+            channels=int(self.intermediate_size),
+            kernel_size=int(self.conv_kernel_size),
+            bias=self.use_conv_bias,
+            padding=int(self.conv_kernel_size - 1)
+        )
+
+        self.x_proj = nn.Linear(self.intermediate_size, self.time_step_rank + 2 * self.ssm_state_size, bias=False)
+        self.dt_proj = nn.Linear(self.time_step_rank, self.intermediate_size, bias=True)
+
+        dt_init_std = args.dt_rank**-0.5 * args.state_size
+        if args.time_step_init_scheme == "constant":
+            self.dt_proj.weight = dt_init_std * mx.ones_like(self.dt_proj.weight)
+        elif args.time_step_init_scheme == "random":
+            self.dt_proj.weight = mx.random.uniform(-dt_init_std, dt_init_std, self.dt_proj.weight.shape)
+        else:
+            raise NotImplementedError
+
+        dt = clamp(mx.exp(
+            mx.random.uniform(shape=[args.intermediate_size]) * (math.log(args.time_step_max) - math.log(args.time_step_min)) + math.log(args.time_step_min)
+        ), min=args.time_step_floor)
+        inv_dt = dt + mx.log1p(-mx.exp(-dt))
+        self.dt_proj.bias = inv_dt
+
+        A = mx.repeat(mx.arange(1., self.ssm_state_size + 1.).reshape([1, self.ssm_state_size]), repeats=self.intermediate_size, axis=0)
+        self.A_log = mx.log(A)
+        self.D = mx.ones([self.intermediate_size])
+
+        self.out_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=args.use_bias)
+
+    def ssm(self, x, h):
+        A = -mx.exp(self.A_log)  # (ED, N)
+        D = self.D
+
+        deltaBC = self.x_proj(x)  # (B, dt_rank+2*N)
+
+        delta, B, C = mx.split(deltaBC, indices_or_sections=[self.time_step_rank, self.time_step_rank+self.ssm_state_size], axis=-1)  # (B, dt_rank), (B, N), (B, N)
+        delta = nn.softplus(self.dt_proj(delta))  # (B, ED)
+
+        deltaA = mx.exp(mx.expand_dims(delta, -1) * A)  # (B, ED, N)
+        deltaB = mx.expand_dims(delta, -1) * mx.expand_dims(B, 1)  # (B, ED, N)
+
+        BX = deltaB * mx.expand_dims(x, -1)  # (B, ED, N)
+
+        if h is None:
+            h = mx.zeros([x.shape[0], self.intermediate_size, self.ssm_state_size])  # (B, ED, N)
+
+        h = deltaA * h + BX  # (B, ED, N)
+
+        y = mx.sum(h * mx.expand_dims(C, 1), axis=-1)  # (B, ED)
+
+        y = y + D * x
+        return y, h
+    
+    def __call__(self, x, cache):
+        h, inputs = cache
+        
+        x, z = self.in_proj(x).split(indices_or_sections=2, axis=-1)
+        
+        # x is now (B, L, C), we need (B, C, L) for conv1d
+        x_cache = x.transpose(0, 2, 1)
```

Comment:
Convolutions in MLX are channels last, so you shouldn't need to transpose anything (just make sure to transpose the weights when you convert the model, kind of like [here](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/models/recurrent_gemma.py#L510-L517))